src/conduits/RwaOutputConduit3.sol

```diff
@@ -235,27 +278,63 @@ contract RwaOutputConduit3 {
      * @dev `msg.sender` must have been `mate`d and `to` must be setted.
      */
     function push() external isMate {
+        _doPush(dai.balanceOf(address(this)), 0);
+    }
+
+    /**
+     * @notice Method to swap specific amount of DAI contract balance to GEM through PSM and push it to the recipient address.
+     * @dev `msg.sender` must have been `mate`d and `to` must be setted.
+     * @param wad Dai amount
+     */
+    function push(uint256 wad) external isMate {
+        _doPush(wad, gem.balanceOf(address(this)));
+    }
+
+    /**
+     * @notice Flushes out any DAI balance to `quitTo` address.
+     * @dev `msg.sender` must first receive push acess through mate().
+     */
+    function quit() external isMate {
+        _doQuit(dai.balanceOf(address(this)));
+    }
+
+    /**
+     * @notice Flushes out specific amount of DAI balance to `quitTo` address.
+     * @dev `msg.sender` must first receive push acess through mate().
+     * @param wad Dai amount
+     */
+    function quit(uint256 wad) external isMate {
+        _doQuit(wad);
+    }
+
+    /**
+     * @notice Internal method to swap specific amount of DAI contract balance to GEM through PSM and push it to the recipient address.
+     * @param wad Dai amount
+     * @param prevGemBalance Previouse GEM balance used to track exect maount of GEM swapped for DAI from PSM (Set 0 if you want to get all outstanding GEM balance)
+     */
+    function _doPush(uint256 wad, uint256 prevGemBalance) internal {
         require(to != address(0), "RwaOutputConduit3/to-not-picked");
 
-        uint256 balance = dai.balanceOf(address(this));
-        uint256 gemAmount = balance / toGemConversionFactor;
+        // We can lose some dust there. For exm: USDC has 6 dec and DAI has 18
+        uint256 gemAmount = wad / toGemConversionFactor;
         require(gemAmount > 0, "RwaOutputConduit3/insufficient-swap-gem-amount");
 
         psm.buyGem(address(this), gemAmount);
 
         uint256 gemBalance = gem.balanceOf(address(this));
-        gem.transfer(to, gemBalance);
+        uint256 gemPushAmt = gemBalance - prevGemBalance;
```

Comment:
We should probably use the safe `sub()` method here.

While it's probably fine to assume that after calling `psm.buyGem` the current gem balance is greater than the balance before, technically `prevGemBalance` is an unbounded argument of this function and this expression can still underflow.

Reply:
Done